### PR TITLE
pipeline - add support to multi processes

### PR DESCRIPTION
**Pull request by raghaven447**

### CODE REVIEW

File: daemon/consumerd/gaeguli-consumer.c [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+#include <wait.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <errno.h>
+#include <unistd.h>
+#include <syslog.h>
+#include <string.h>
+#include <poll.h>
```
Comment (justinjoy) [inline daemon/consumerd/gaeguli-consumer.c:36]: Without glib, we might not think of starting this project.
Sometimes, it would be simpler to write the codes from primitive functions than using abstract APIs, but it has less advantage when we think of maintenance.

GLib provides GSpwaning as the alternatives of `fork()` and `exec()`. That way will make us focus on the core rather than considering compatibility.

File: daemon/consumerd/test/gaeguli-consumer-client.c [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+static void
+_display_menu (void)
+{
+  printf ("********* MENU *********\n\n");
+  printf ("Enter 1 to create Target pipeline\n");
+  printf ("Enter 2 to destroy Target pipeline\n");
+  printf ("Enter 3 to exit\n");
+  printf ("************************\n");
+}
```
Comment (justinjoy) [inline daemon/consumerd/test/gaeguli-consumer-client.c:63]: I'd like to keep the consistency of using `print()` things.
As I mentioned above, this project is being built on top of GLib. Thus, I recommend using macros or functions from GLib as much as possible.

File: daemon/consumerd/test/gaeguli-consumer-client.c [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+_handle_create_target (GaeguliPipeline * pipeline,
+    const gchar * uri, const gchar * username, guint input_node_id,
+    guint output_node_id, GaeguliTargetType type)
+{
+  GaeguliTarget *target = NULL;
+  g_autoptr (GError) error = NULL;
+
+  switch (type) {
+    case GAEGULI_TARGET_TYPE_SRT:{
+      g_print ("Client:: Invoking gaeguli_pipeline_add_srt_target() "
```
Comment (justinjoy) [inline daemon/consumerd/test/gaeguli-consumer-client.c:75]: Perhaps, it can be debugging messages. Can't we use `g_info()` or `g_message()`?

File: daemon/consumerd/gaeguli-consumer.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  int fd_idx = 0;
+  int resfd = 0;
+
+  gst_init (NULL, NULL);        /* Init the gst ebvironment here */
+
+  while (1) {
+    resfd = poll (fds, numfds, 10000);
+    if (resfd < 0) {
+      if (errno == EINTR) {
+        syslog (LOG_ERR, "Got an interrupt. error (%s).", strerror (errno));
```
Comment (justinjoy) [inline daemon/consumerd/gaeguli-consumer.c:77]: Although `syslog()` is definitely a good tool, I don't think we need to use it directly.
Most debug messages of the others are managed by `g_log()` and we can check the messages by _journalctl_.

File: gaeguli/pipeline.h [hunk 5]
```diff
@@ -65,14 +65,29 @@ GaeguliPipeline        *gaeguli_pipeline_new    (void);
  */
 GaeguliPipeline        *gaeguli_pipeline_new_full
                                                 (GaeguliVideoSource     source,
-                                                 const gchar           *device,
+                                                 guint                  node_id,
```
Comment (justinjoy) [inline gaeguli/pipeline.h:68]: Even though `pipewire` is used internally, I don't think we don't need to expose what we are using to user.
`node_id` is a concept too stick to pipewire, I think.

Instead, `GaeguliPipeline` should maintain the node information internally, and user should be able to create the pipeline by the device name as it is.

File: gaeguli/pipeline.h [hunk 6]
```diff
@@ -65,14 +65,29 @@ GaeguliPipeline        *gaeguli_pipeline_new    (void);
...
+void                    gaeguli_pipeline_create_source
+                                                (GaeguliPipeline * self,
+                                                 GError ** error);
 /**
  * gaeguli_pipeline_add_srt_target:
  * @self: a #GaeguliPipeline object
  * @uri: SRT URI
  * @username: SRT Stream ID User Name identifying this target
+ * @input_node_id: pipewire node id identifying the video source
+ * output_node_id: pipewire node id from which the target gets the video feed
```
Comment (justinjoy) [inline gaeguli/pipeline.h:90]: I think it's too complex if we let users maintain node and connection information. 
I'd prefer providing only pipeline and target to user to prevent programmer errors.

File: gaeguli/types.h [hunk 7]
```diff
@@ -54,6 +54,7 @@ typedef enum {
   GAEGULI_VIDEO_SOURCE_AVFVIDEOSRC,
   GAEGULI_VIDEO_SOURCE_VIDEOTESTSRC,
   GAEGULI_VIDEO_SOURCE_NVARGUSCAMERASRC,
+  GAEGULI_VIDEO_SOURCE_PIPEWIRESRC,
```
Comment (justinjoy) [inline gaeguli/types.h:57]: We should think of what video source means.
According to the implicit norm of `GaeguliVideoSource`, it means a source to get a video stream.

The others except for _pipewiresrc_ tend to fit this definition. However, _pipewiresrc_ here sounds like another video source from pipewire that is not managed by _Gaeguli_. For this reason, we don't need to define pipewiresrc explicitly.

File: gaeguli/types.h [hunk 8]
```diff
@@ -96,6 +97,13 @@ typedef enum {
   GAEGULI_TARGET_STATE_ERROR
 } GaeguliTargetState;
 
+typedef enum {
+  GAEGULI_TARGET_TYPE_UNKNOWN,
+  GAEGULI_TARGET_TYPE_SRT,
+  GAEGULI_TARGET_TYPE_RECORDING,
+  GAEGULI_TARGET_TYPE_IMAGE_CAPTURE
+} GaeguliTargetType;
```
Comment (justinjoy) [inline gaeguli/types.h:105]: Adding new types should be in a separate commit.

File: gaeguli/types.h [hunk 9]
```diff
@@ -121,4 +129,33 @@ typedef enum {
...
+  GAEGULI_CONSUMER_MSG_DESTROY_TARGET
+} GaeguliConsumerMsgType;
+
+typedef enum {
+  GAEGULI_CONSUMER_RSP_UNKNOWN = 0,
+  GAEGULI_CONSUMER_RSP_CREATE_TARGET_SUCCESS,
+  GAEGULI_CONSUMER_RSP_START_TARGET_SUCCESS,
+  GAEGULI_CONSUMER_RSP_DESTROY_TARGET_SUCCESS,
+  GAEGULI_CONSUMER_RSP_FAIL
+} GaeguliConsumerRspType;
```
Comment (justinjoy) [inline gaeguli/types.h:160]: I have two major concerns with these definitions. First, I am wondering if we need to make these enums public.
Second, it looks like _source_ and _consumer_ needs similar message types.
If they are really required in different enums, they should be defined by separated commits.
But, I think it is better to consider simple message types because it looks like internal-used enums.

File: gaeguli/utils/gaeguli_utils.h [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#ifndef __GAEGULI_UTILS_H__
+#define __GAEGULI_UTILS_H__
+
```
Comment (justinjoy) [inline gaeguli/utils/gaeguli_utils.h:21]: This header makes me confusing because it looks like APIs internal use only.

When we discussed multi-process architecture, I expected that _GaeguliPipeline_ and _GaeguliTarget_ would have _provider_ and _consumer_ services respectively. That means managing each service is not exposed to user. 

Does an application like _Gaeul-Source-Agent_ need APIs to manage internal services? If so, I think it is a wrong start.

File: tests/adaptor-demo/adaptor-demo.c [hunk 11]
```diff
@@ -148,7 +148,7 @@ gaeguli_adaptor_demo_on_msg_stream (GaeguliAdaptorDemo * self, JsonObject * msg)
       }
 
       self->target = gaeguli_pipeline_add_srt_target_full (self->pipeline,
-          codec, 2048000, "srt://:7001?mode=listener", NULL, &error);
+          codec, 2048000, "srt://:7001?mode=listener", NULL, 0, 0, &error);
```
Comment (justinjoy) [inline tests/adaptor-demo/adaptor-demo.c:151]: Why is `s/NULL/0` required? To prevent confusion, if it is not an integer type, it is better to stay `NULL`.

**Review by justinjoy — CHANGES_REQUESTED** [review]
Although I don't read deeply, this PR should be revised.

File: gaeguli/types.h (hunk 7 above)
Comment (xhaakon) [reply gaeguli/types.h:57]: Yes adding Pipewire as another potential source of live camera stream (alongside v4l2 etc.) and splitting GaeguliPipeline into multiple processes that internally utilize Pipewire to pass video frames between each other are two very different features.

If you implemented both, @raghaven447, please move the code that deals with Pipewire stream source (that means pipewiresrc providing stream directly from /dev/video...) into a separate MR. Here, let's focus only on making Gaeguli multiprocess, where one process acts as the stream source (something like `v4l2src device=/dev/video0 ! decodebin ! pipewiresink mode=provide`) and the rest of the processes are consumers that encode the stream and either send it over the network or store it locally (e.g. `pipewiresrc ! x264enc ! srtsink`). In such scenario, pipewire should be virtually invisible to the user of Gaeguli API.

File: gaeguli/pipeline.c [hunk 12]
```diff
@@ -475,78 +452,79 @@ gaeguli_pipeline_new_full (GaeguliVideoSource source, const gchar * device,
...
+          gaeguli_configure_client_socket
+          (DEFAULT_SOURCE_PROVIDER_CLIENT_SOCK_PATH)) < 0) {
+    syslog (LOG_ERR, "Failed to create client socket");
+    return NULL;
+  }
 
-    if (!gst_tag_exists (tag_name)) {
-      g_warning ("Unknown tag %s", tag_name);
-      goto next;
+  /* send IPC msg to provider daemon */
```
Comment (xhaakon) [inline gaeguli/pipeline.c:487]: To me, this architecture with provider and consumer daemons that in turn spawn another child processes seems unnecessarily complicated. I think we don't need `providerd` and `consumerd` middle-men. The Gaeguli API client application that called `gaeguli_pipeline_new_full ()` IS the manager of all child processes belonging to that pipeline.

I think what `gaeguli_pipeline_new_full ()` should do is to launch a child process that runs the video source pipeline (`v4l2src ! decodebin ! pipewiresink`) and retrieve `node_id` of the new `pipewiresink` from this _VSRC Child_ process. Allocating `node_id` by the client application and passing it to _VSRC Child_ is also an option; point is both must know what `node_id` value is.

Subsequently, each `gaeguli_pipeline_add_srt_target ()` call should spawn a new _Consumer Child_ (`pipewiresrc ! encoder ! srtsink`), passing it _VSRC Child_'s `node_id`.

`gaeguli_pipeline_remove_target ()` terminates its respective _Consumer Child_.
`gaeguli_pipeline_stop ()` terminates _VSRC Child_ and all _Consumer Child_ instances.

The client application monitors the state of its children and should some terminate unexpectedly, `GaeguliPipeline` will launch it again (so it has to remember the parameters the child was created with for potential resurrection).

In this setup, I imagine the needs for IPC between `GaeguliPipeline` and its children besides passing the initial parameters will be minimal if any.

@justinjoy Does this look like what you want?

File: gaeguli/pipeline.h (hunk 6 above)
Comment (xhaakon) [reply gaeguli/pipeline.h:90]: What is the second `node_id` anyway? As I understand it, `GaeguliTarget` only needs to know the `node_id` of `pipewiresink` that appears in `GAEGULI_PIPELINE_VSRC_STR`.

**Review by xhaakon — CHANGES_REQUESTED** [review]
I'm a bit concerned about reviewing such huge monolithic commits. As it appears there is still some confusion about the direction this feature should head towards, I think the key to success is proceeding in smaller steps and getting more feedback after each of them.

@raghaven447 Let's take a few steps back and split this into more palatable packages. As the first step, I think you could provide a MR that still has `GaeguliPipeline` all in one process and _only_ replaces `tee` in `GAEGULI_PIPELINE_VSRC_STR` with `pipewiresink` and `pipewiresrc`s.

Then I think we could add management of child processes - let every `gaeguli_pipeline_new ()` create a child process that will just sit idly during the pipeline's lifetime and terminate on `gaeguli_pipeline_stop()`. Same for `gaeguli_pipeline_add_srt_target()` and `gaeguli_pipeline_remove_target()`. In this step you can also implement some watchdog for child processes that will relaunch one if it crashes.

The next step could move the pipelines that were split in the first MR but kept running in the main process into their respective children.

After that, we shall see :)

Let's check and discuss the code created in each step together with @justinjoy before moving forward.

File: gaeguli/pipeline.c (hunk 12 above)
Comment (justinjoy) [reply gaeguli/pipeline.c:487]: Sorry for the late reply, but I agree. It looks like a simple and clear approach.

**Comment by raghaven447** [comment]
Closing this PR as the refactored changes would be submitted in other PRs.
